### ISSUE TITLE
SCM: Set placeholder text in SourceControlInputBox

### ISSUE THREAD

**Issue by stef-levesque** [Microsoft/vscode#24090]
- VSCode Version: Code 1.11.1 (d9484d12b38879b7f4cdd1150efeb2fd2c1fbf39, 2017-04-06T14:00:34.044Z)
- OS Version: Darwin x64 16.5.0
- Extensions: 

|Extension|Author|Version|
|---|---|---|
|[perforce](https://github.com/stef-levesque/vscode-perforce)|slevesque|1.1.0|;

---

I would need a way to change the placeholder text for scm.inputBox .
I intent to use the box to display and update changelist description, and so should display:
`Description (press Cmd+Enter to create changelist)`
as a placeholder when the default changelist is selected.

**Comment by stef-levesque** [comment]
Not exactly orthodoxe, but...

```patch
From 595a891406d3d390eac12ffdfc6b9d7bd7010910 Mon Sep 17 00:00:00 2001
From: Stephane Levesque <stef.levesque@gmail.com>
Date: Fri, 7 Apr 2017 19:09:17 -0400
Subject: [PATCH] :sparkles: Optional placeholder for the Source Control input

Fixes #24090
---
 src/vs/vscode.d.ts                                        |  8 ++++++++
 src/vs/workbench/api/node/extHost.protocol.ts             |  1 +
 src/vs/workbench/api/node/extHostSCM.ts                   | 11 +++++++++++
 src/vs/workbench/api/node/mainThreadSCM.ts                |  8 ++++++++
 src/vs/workbench/parts/scm/electron-browser/scmViewlet.ts | 13 ++++++++++---
 src/vs/workbench/services/scm/common/scm.ts               |  2 ++
 6 files changed, 40 insertions(+), 3 deletions(-)

diff --git a/src/vs/vscode.d.ts b/src/vs/vscode.d.ts
index 70b052ae6..1a70ecc6f 100644
--- a/src/vs/vscode.d.ts
+++ b/src/vs/vscode.d.ts
@@ -4694,6 +4694,14 @@ declare module 'vscode' {
 		commitTemplate?: string;
 
 		/**
+		 * Optional commit placeholder string.
+		 *
+		 * The Source Control will show this as place holder
+		 * to explain the user the purpose of the input
+		 */
+		commitPlaceholder?: string;
+
+		/**
 		 * Optional accept input command.
 		 *
 		 * This command will be invoked when the user accepts the value
diff --git a/src/vs/workbench/api/node/extHost.protocol.ts b/src/vs/workbench/api/node/extHost.protocol.ts
index d1542e816..0cb5791f2 100644
--- a/src/vs/workbench/api/node/extHost.protocol.ts
+++ b/src/vs/workbench/api/node/extHost.protocol.ts
@@ -254,6 +254,7 @@ export interface SCMProviderFeatures {
 	hasQuickDiffProvider?: boolean;
 	count?: number;
 	commitTemplate?: string;
+	commitPlaceholder?: string;
 	acceptInputCommand?: modes.Command;
 	statusBarCommands?: modes.Command[];
 }
diff --git a/src/vs/workbench/api/node/extHostSCM.ts b/src/vs/workbench/api/node/extHostSCM.ts
index 6b621f709..9612f0053 100644
--- a/src/vs/workbench/api/node/extHostSCM.ts
+++ b/src/vs/workbench/api/node/extHostSCM.ts
@@ -192,6 +192,17 @@ class ExtHostSourceControl implements vscode.SourceControl {
 		this._proxy.$updateSourceControl(this._handle, { commitTemplate });
 	}
 
+	private _commitPlaceholder: string | undefined = undefined;
+
+	get commitPlaceholder(): string | undefined {
+		return this._commitPlaceholder;
+	}
+
+	set commitPlaceholder(commitPlaceholder: string | undefined) {
+		this.commitPlaceholder = commitPlaceholder;
+		this._proxy.$updateSourceControl(this._handle, { commitPlaceholder });
+	}
+
 	private _acceptInputCommand: vscode.Command | undefined = undefined;
 
 	get acceptInputCommand(): vscode.Command | undefined {
diff --git a/src/vs/workbench/api/node/mainThreadSCM.ts b/src/vs/workbench/api/node/mainThreadSCM.ts
index 76f7f976e..aee74ab46 100644
--- a/src/vs/workbench/api/node/mainThreadSCM.ts
+++ b/src/vs/workbench/api/node/mainThreadSCM.ts
@@ -79,12 +79,16 @@ class MainThreadSCMProvider implements ISCMProvider {
 	get id(): string { return this._id; }
 
 	get commitTemplate(): string | undefined { return this.features.commitTemplate; }
+	get commitPlaceholder(): string | undefined { return this.features.commitPlaceholder; }
 	get acceptInputCommand(): Command | undefined { return this.features.acceptInputCommand; }
 	get statusBarCommands(): Command[] | undefined { return this.features.statusBarCommands; }
 
 	private _onDidChangeCommitTemplate = new Emitter<string>();
 	get onDidChangeCommitTemplate(): Event<string> { return this._onDidChangeCommitTemplate.event; }
 
+	private _onDidChangeCommitPlaceholder = new Emitter<string>();
+	get onDidChangeCommitPlaceholder(): Event<string> { return this._onDidChangeCommitPlaceholder.event; }
+
 	private _count: number | undefined = undefined;
 	get count(): number | undefined { return this._count; }
 
@@ -104,6 +108,10 @@ class MainThreadSCMProvider implements ISCMProvider {
 		if (typeof features.commitTemplate !== 'undefined') {
 			this._onDidChangeCommitTemplate.fire(this.commitTemplate);
 		}
+
+		if (typeof features.commitPlaceholder !== 'undefined') {
+			this._onDidChangeCommitPlaceholder.fire(this.commitPlaceholder);
+		}
 	}
 
 	$registerGroup(handle: number, id: string, label: string): void {
diff --git a/src/vs/workbench/parts/scm/electron-browser/scmViewlet.ts b/src/vs/workbench/parts/scm/electron-browser/scmViewlet.ts
index 375539eef..d4645679a 100644
--- a/src/vs/workbench/parts/scm/electron-browser/scmViewlet.ts
+++ b/src/vs/workbench/parts/scm/electron-browser/scmViewlet.ts
@@ -253,6 +253,10 @@ export class SCMViewlet extends Viewlet {
 				disposables.push(activeProvider.onDidChangeCommitTemplate(this.updateInputBox, this));
 			}
 
+			if (activeProvider.onDidChangeCommitPlaceholder) {
+				disposables.push(activeProvider.onDidChangeCommitPlaceholder(this.updateInputBox, this));
+			}
+
 			this.providerChangeDisposable = combinedDisposable(disposables);
 		} else {
 			this.providerChangeDisposable = EmptyDisposable;
@@ -354,11 +358,14 @@ export class SCMViewlet extends Viewlet {
 			return;
 		}
 
-		if (typeof this.activeProvider.commitTemplate === 'undefined') {
-			return;
+		if (typeof this.activeProvider.commitTemplate !== 'undefined') {
+			this.inputBox.value = this.activeProvider.commitTemplate;
+		}
+
+		if (typeof this.activeProvider.commitPlaceholder !== 'undefined') {
+			this.inputBox.setPlaceHolder(this.activeProvider.commitPlaceholder);
 		}
 
-		this.inputBox.value = this.activeProvider.commitTemplate;
 	}
 
 	layout(dimension: Dimension = this.cachedDimension): void {
diff --git a/src/vs/workbench/services/scm/common/scm.ts b/src/vs/workbench/services/scm/common/scm.ts
index e6eaaf95c..5ddaf15a5 100644
--- a/src/vs/workbench/services/scm/common/scm.ts
+++ b/src/vs/workbench/services/scm/common/scm.ts
@@ -47,6 +47,8 @@ export interface ISCMProvider extends IDisposable {
 	readonly count?: number;
 	readonly commitTemplate?: string;
 	readonly onDidChangeCommitTemplate?: Event<string>;
+	readonly commitPlaceholder?: string;
+	readonly onDidChangeCommitPlaceholder?: Event<string>;
 	readonly acceptInputCommand?: Command;
 	readonly statusBarCommands?: Command[];
 
-- 
2.11.0 (Apple Git-81)

```

**Comment by joaomoreno** [comment]
Fixed by https://github.com/Microsoft/vscode/pull/29822
